### DISCUSSION TITLE
Feature Request: Give sections an option to route to a controller not just a template

### DISCUSSION THREAD

**Discussion by tjdraper** [craftcms/cms#8505] [Ideas] [answered]
With Craft 3 we can have a custom module for site specific features, controllers, services, twig extensions, you name it. And it's AAAAAMAZING. However, one thing that's bugging me is this: I've worked on (and am working on currently) sites where every page needs to run through custom controllers to make sure custom variables are set, custom logic is run etc. I've jumped through elaborate hoops to make this work in Craft 2. But in Craft 3, it seems to me there's an opportunity to allow routing to a custom Yii controller as an option and not just a template. If we could specify the controller and method a section should run for the entry's URL in the same way we can specify a Yii controller in a route (for instance `dev/locations/entry` in the case of a site I'm working on right now, that would be super duper fantastic.

So basically, as a feature request, anywhere we can tell Craft in the CMS to route a URI to a template, also give us the option to route to a controller action.

**Comment by tjdraper** [comment]
Meant to say also that the entry that's normally pre-loaded as a variable available in Twig could be an incoming route parameter or attached to the route object in some way.

**Comment by brandonkelly** [comment, marked as answer]
I’ve thought about this, but unsure whether enough people would want it, to be worth the added UI complexity.

You can take over elements’ routing though, using the [`EVENT_SET_ROUTE`](https://docs.craftcms.com/api/v3/craft-base-element.html#EVENT_SET_ROUTE-detail) event.

```php
use craft\base\Element;
use craft\events\SetElementRouteEvent;
use yii\base\Event;

Event::on(Element::class, Element::EVENT_SET_ROUTE, function(SetElementRouteEvent $e) {
    /** @var Element $element */
    $element = $e->sender;

    $params = [
        'element' => $element,
    ];
    $e->route = ['dev/locations/entry', $params];
});
```

**Comment by tjdraper** [comment]
Very helpful, thanks! The only additional argument I'll make in favor of the feature request is perhaps maybe having the route start with a keyword Craft can look for, no additional UI needed. Maybe something like `actionRoute/dev/locations/entry`. Otherwise, manually doing something like this with that event is probably fine.

**Comment by brandonkelly** [comment]
Not a bad idea!

**Comment by brandonkelly** [comment]
Played around with it. Issue with supporting the template path being set with something `action:` is, the column is _called_ “Template”. We’r really need to rename it to “Route” if we support anything besides template paths. And if we rename it to “Route”, it would be pretty awkward that routes defined here default to template paths and you need to specify `action:` if you want it to go to a controller action; but routes defined in config/routes.php default to controller actions, and if you want it to go to a template, you have to specify that with `['template' => 'template/path']`. So I think when we add support for action paths here, we just need both types of routes to explicitly define what they are.

Probably something like this:

<img width="297" alt="screen shot 2018-02-19 at 4 02 03 pm" src="https://user-images.githubusercontent.com/47792/36401789-ac70d08e-158e-11e8-82eb-6f8acf174b72.png">

**Comment by carlcs** [comment]
Don't forget about #1489 when thinking about updates to routes. 😉

**Comment by brandonkelly** [comment]
Good point @carlcs.

**Comment by gijsstegehuis** [comment]
@brandonkelly given your piece of code:
`$e->route = ['dev/locations/entry', $params];`
Where would `$params` end up?

I routed it onto a module controller, but the params aren't being set as controller action arguments.

**Comment by gijsstegehuis** [comment]
Nevermind that @brandonkelly.

UrlManager will do the trick:
`Craft::$app->getUrlManager()->getRouteParams();`

**Comment by brandonkelly** [comment]
@gijsstegehuis Any parameters that match the names of your controller action arguments will be passed to those. So you could do this:

```php
public function actionEntry(\craft\base\ElementInterface $element)
{
    // $element will be whatever $params['element'] was in the route
}
```

**Comment by davidwebca** [comment]
Hi from the future! I'm currently searching for a way to keep Craft's core behaviour intact by setting templates in entries sections in the CP... but intercepting those templates rendering with a Controller to be able to structure my code better and avoid querying data inside twig views. Something akin to View Composers in Laravel? One class per view that would allow to structure code with OOP, execute my queries before rendering the template and maybe execute some custom sql before passing data down... 🤔  There are a lot of solutions to this problem, but I'm mostly searching for the best ideal way to do this and I stumbled onto this issue. 

Any chance that theres a clean solution in 2022?

**Reply by tjdraper** [reply]
This may be too much take over of Craft for you, but I wrote this (for Craft 3, no idea if it will work with Craft 4 as I haven't had occasion to do Craft work lately and the project I wrote this for hasn't been upgraded to Craft 4 yet) https://github.com/buzzingpixel/craft-slim-bridge

I made it public because I figured if anyone wants to use it, they're welcome to. I probably can't offer much support for it. Though I actively use it on a project.

Essentially, when I'm working with Craft and this plugin, I route EVERYTHING on the front-end to a PHP class, and that class can render a template (or whatever it wants).

As an example, here's how Slim Bridge picks up a route from a page structure and routes it to a class in my project.

<img width="1127" alt="Screen Shot 2022-11-17 at 2 40 51 PM" src="https://user-images.githubusercontent.com/3803475/202555190-06fd5a6d-9066-4a40-b043-3c80cbc31d4e.png">

Here's an example of my `PageAction` class

<img width="655" alt="image" src="https://user-images.githubusercontent.com/3803475/202555490-d3fa9237-661c-403b-a279-13080d6603a5.png">

Here's an example of me creating a route that isn't routed through the Craft CP:

<img width="707" alt="Screen Shot 2022-11-17 at 2 47 27 PM" src="https://user-images.githubusercontent.com/3803475/202556481-f100d403-f964-4332-ba6a-6123e54b1c87.png">

<img width="622" alt="Screen Shot 2022-11-17 at 2 48 34 PM" src="https://user-images.githubusercontent.com/3803475/202556680-9661c409-0226-401d-ba87-f100aaeda421.png">

<img width="678" alt="Screen Shot 2022-11-17 at 2 49 03 PM" src="https://user-images.githubusercontent.com/3803475/202556794-be059112-9931-4662-856d-27ad22452e46.png">

I realize this way of working might not be for everyone, but I'm a PHP developer, I enjoy working with Slim and have apps and sites I manage using the Slim framework, I enjoy PSR standard interfaces, and I enjoy a proper separation of concerns and not putting everything in Twig templates. So, if this might scratch your itch, you're welcome to it.

**Reply by tjdraper** [reply]
BTW, Slim Bridge for CraftCMS uses, essentially, this method from Brandon: https://github.com/craftcms/cms/discussions/8505#discussioncomment-901735

https://github.com/buzzingpixel/craft-slim-bridge/blob/master/src/ElementSetRoute/SetRouteFromParsed/SetRoute.php

**Reply by davidwebca** [reply]
Very interesting and thorough answer, thank you so much! 

You're absolutely right: it is mostly for separation of concerns. I've been working with many PHP frameworks over the years and this is the one thing I'd like to know before starting my next project with Craft. 

I'll consider this for sure, but if anyone else has tackled this in "pure" Craft I'd be interested to hear more.
